src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -644,26 +646,37 @@ private void Dispose()
                 {
                     if (!_noImplicitTransforms)
                     {
+                        // given the type, fetch the implicit transforms for that type, we cache this in 
+                        // one of two places, and if we don't find it we generate it from the type (MakeImplicitTransforms)
                         Type argType = args.GetType();
-                        if (_expectedArgType != argType)
+                        TransformSpec implicitTransforms;
+                        if (argType == _firstImplicitFieldsType && _firstImplicitFieldsTypeTransforms != null)
                         {
-                            // Figure out the default properties to send on to EventSource.  These are all string or primitive properties.  
-                            _implicitTransforms = null;
-                            TransformSpec newSerializableArgs = null;
-                            TypeInfo curTypeInfo = argType.GetTypeInfo();
-                            foreach (var property in curTypeInfo.DeclaredProperties)
-                            {
-                                var propertyType = property.PropertyType;
-                                newSerializableArgs = new TransformSpec(property.Name, 0, property.Name.Length, newSerializableArgs);
-                            }
-                            _expectedArgType = argType;
-                            _implicitTransforms = Reverse(newSerializableArgs);
+                            // This is the hoped for case.  We hit the singleton cache that we have already set.  
+                            implicitTransforms = _firstImplicitFieldsTypeTransforms;
+                        }
+                        else if (_firstImplicitFieldsType == null)
+                        {
+                            // The singleton cache slot is empty, so create the implicit transforms for that type and cache it.  
+                            implicitTransforms = MakeImplicitTransforms(argType);
+                            // Threads race to claim this slot but only one wins, and gets the right to update _firstImplicitFieldsTypeTransforms. 
+                            // We never modify it after that.   
+                            if (implicitTransforms != null && Interlocked.CompareExchange(ref _firstImplicitFieldsType, argType, null) == null)
+                                _firstImplicitFieldsTypeTransforms = implicitTransforms;
+                        }
+                        else
+                        {
+                            // This should only happen when you are wildcarding your events (reasonably rare).   In that case you will probably need many types
```

Comment:
It seems like this could also happen if there's a race condition where two threads race to initialize the single-item cache with the same thing:
- Thread 1 runs and in the "else if (_firstImplicitFieldsType == null)" branch above successfully interlocked exchanges to set _firstImplicitFieldsType to argType… but hasn't yet set _firstImplicitFieldsTypeTransforms to implicitTransforms.
- Thread 2 runs, sees that argType == _firstImplicitFieldsType but _firstImplicitFieldsTypeTransforms is still null, so it falls through both if and else if blocks above and ends up here, initializing the concurrent dictionary even though it's not needed.

Seems like both _firstImplicitFieldsType and _firstImplicitFieldsTypeTransforms should be part of the same object, with just a single field for it on the event source, so that they can both be swapped in atomically together.

Reply:
I was tolerating races before, but I agree that it is better to avoid them, so I have taken your suggestion.